contracts/OtokenFactory.sol

```diff
@@ -0,0 +1,185 @@
+pragma solidity =0.6.10;
+
+import {OtokenSpawner} from "./OtokenSpawner.sol";
+import {AddressBookInterface} from "./interfaces/AddressBookInterface.sol";
+import {OtokenInterface} from "./interfaces/OtokenInterface.sol";
+import {WhitelistInterface} from "./interfaces/WhitelistInterface.sol";
+
+/**
+ * SPDX-License-Identifier: UNLICENSED
+ * @title A factory for opyn tokens
+ * @author Opyn
+ * @notice Create new otokens and keep track of all created tokens.
+ * @dev Calculate contract address before each creation with CREATE2
+ * and deploy eip-1167 minimal proxies for otoken logic contract.
+ */
+contract OtokenFactory is OtokenSpawner {
+    /// @notice The Opyn AddressBook contract that records addresses of whitelist module and otoken impl address. */
+    address public addressBook;
+
+    /// @dev An array of all created otokens */
+    address[] private _otokens;
```

Comment:
@haythem96 are we for variables using _name convention or only for functions?

Reply:
Also do we know how many oTokens are created? Like can we access the length of the array externally? (do we need that data?

Reply:
Actually I'm thinking about can we just have a counter and a public mapping. Returning the whole array could be problematic when we have lots of otokens in the future. (Similar to the `getVaultOwners` function we have in v1...)

Reply:
We can have a public array and a counter, and a getter function that return the counter, so off-chain it just need to loop from 0 to counter-1, similar to here https://github.com/opynfinance/OptionsProtocol/blob/master/contracts/OptionsFactory.sol#L102

Reply:
@haythem96 that's a good idea, it's how Uniswap v2 handles it too.

https://github.com/Uniswap/uniswap-v2-core/blob/4dd59067c76dea4a0e8e4bfdda41877a6b16dedc/contracts/UniswapV2Factory.sol#L19

Reply:
> 1. Will you have integration tests to test how this works with whitelist module?

We will need addressBook ready to run real integration tests on all of them